parity-clib/src/python.rs

```diff
@@ -0,0 +1,219 @@
+			unsafe {
+				parity_config_destroy(self.inner);
+			}
+		}
+	}
+}
+
+/// Handle to a Parity instance created with parity_start
+#[pyclass]
+struct ParityHandle {
+	inner: *mut c_void,
+}
+
+impl Drop for ParityHandle {
+	fn drop(&mut self) {
+		if self.inner != ptr::null_mut() {
+			unsafe {
+				parity_destroy(self.inner);
+			}
+		}
+	}
+}
+
+/// Handle to a subscription created with parity_ws_worker
+#[pyclass]
+struct SubscriptionHandle {
+	inner: *const c_void,
+}
+
+#[pymethods]
+impl SubscriptionHandle {
+	fn unsubscribe(&mut self) {
+		if self.inner != ptr::null_mut() {
+			unsafe {
+				parity_unsubscribe_ws(self.inner);
+			}
+
+			self.inner = ptr::null_mut();
+		}
+	}
+}
+
+impl Drop for SubscriptionHandle {
+	fn drop(&mut self) {
+		self.unsubscribe()
+	}
+}
+
+/// A Python callable which accepts a String value representing a RPC result or subscription event
+struct PythonCallback(PyObject);
+
+impl PythonCallback {
+	fn new(callback: PyObject) -> Self {
+		Self(callback)
+	}
+}
+
+impl Callback for PythonCallback {
+	fn call(&self, msg: &str) {
+		// Assume this is called from Rust, acquire GIL
+		let gil = Python::acquire_gil();
+		let py = gil.python();
+
+		// Run our callback, and re-raise any exceptions that occur
+		if let Err(e) = self.0.call1(py, msg) {
+			e.restore(py);
+		}
+	}
+}
+
+/// Generate a Parity config given a list of CLI options
+#[pyfunction]
+fn config_from_cli(_py: Python<'_>, cli: &PyList) -> PyResult<ConfigHandle> {
+	let cli_len = cli.len();
+
+	let opts_strings = cli
+		.into_iter()
+		.map(|o| o.extract())
+		.collect::<PyResult<Vec<String>>>()?;
+
+	let opts = opts_strings
+		.iter()
+		.map(|s: &String| s.as_ptr() as *const i8)
+		.collect::<Vec<_>>();
+	let opts_lens = opts_strings
+		.iter()
+		.map(|s| s.as_bytes().len())
+		.collect::<Vec<_>>();
+
+	let mut out = ptr::null_mut();
+	unsafe {
+		match parity_config_from_cli(opts.as_ptr(), opts_lens.as_ptr(), cli_len, &mut out) {
+			0 => Ok(ConfigHandle { inner: out }),
+			_ => Err(ValueError::py_err("failed to create config object")),
+		}
+	}
+}
+
+/// Create a running Parity instance given a config
+#[pyfunction]
+unsafe fn build(
+	_py: Python<'_>,
+	config: &mut ConfigHandle,
+	logger_mode: &str,
+	logger_file: &str,
+) -> PyResult<ParityHandle> {
+	let mut params = ParityParams {
+		configuration: config.inner,
+		..mem::zeroed()
+	};
+
+	parity_set_logger(
+		logger_mode.as_ptr(),
+		logger_mode.as_bytes().len(),
+		logger_file.as_ptr(),
+		logger_file.as_bytes().len(),
+		&mut params.logger,
+	);
+
+	let mut out = ptr::null_mut();
+	let ret = match parity_start(&params, &mut out) {
+		0 => Ok(ParityHandle { inner: out }),
+		_ => Err(RuntimeError::py_err("failed to start Parity")),
+	};
+
+	// Ensure we don't double-free config
+	config.inner = ptr::null_mut();
+	ret
+}
+
+/// Perform a RPC query against a running Parity instance, invoking callback on completion
+#[pyfunction]
+fn rpc_query(
+	_py: Python<'_>,
+	parity: &ParityHandle,
```

Comment:
how is it ensured that this function can't end up with `undefined behavior`?

Reply:
Need some help here, what guarantees should I be ensuring? Is the concern about the behavior of the callback?

Depending on response java code may also need addressing.

Reply:
Sorry, if I my comment wasn't clear enough but what I meant was how is it ensured that this function is memory safe (aliasing, dangling pointer or similar)?

My only concern is that this function is not marked as unsafe, can you explain why?